src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -425,6 +426,23 @@ private async ValueTask<ProtocolToken> ReceiveBlobAsync<TIOAdapter>(TIOAdapter a
                     {
                         // SNI if it exist. Even if we could not parse the hello, we can fall-back to default certificate.
                         _sslAuthenticationOptions!.TargetHost = _lastFrame.TargetName;
+
+                        if (_sslAuthenticationOptions.ServerOptionDelegate != null)
+                        {
+                            ValueTask<SslServerAuthenticationOptions> t =
+                                _sslAuthenticationOptions.ServerOptionDelegate(this, new SslClientHelloInfo(_lastFrame.TargetName, _lastFrame.SupportedVersions),
+                                                                            _sslAuthenticationOptions.UserState, adapter.CancellationToken);
+
+                            if (t.IsCompletedSuccessfully)
+                            {
+                                _sslAuthenticationOptions.UpdateOptions(t.Result);
+                            }
+                            else
+                            {
+                                SslServerAuthenticationOptions userOptions = await t.ConfigureAwait(false);
+                                _sslAuthenticationOptions.UpdateOptions(userOptions);
+                            }
+                        }
```

Comment:
```suggestion
                            SslServerAuthenticationOptions userOptions =
                                await _sslAuthenticationOptions.ServerOptionDelegate(this, new SslClientHelloInfo(_lastFrame.TargetName, _lastFrame.SupportedVersions),
                                                                            _sslAuthenticationOptions.UserState, adapter.CancellationToken).ConfigureAwait(false);
                            _sslAuthenticationOptions.UpdateOptions(userOptions);
                        }
```

Reply:
I was trying to avoid `await` if task finished synchronously. I assume this is cheap for value task? It certainly simplifies the code.

Reply:
It's cheap.  Just await it.  The main time you might consider doing something like what you did is if there's something more expensive you want to delay doing until you know it didn't complete synchronously, e.g.
```C#
ValueTask<int> vt = SomeOperation();
int result;
if (vt.IsCompletedSynchronously)
{
    result = vt.Result;
}
else
{
    
    using (cancellationToken.Register(...))
    {
        result = await vt;
    }
}
```